### DISCUSSION TITLE
Catching tool errors

### DISCUSSION THREAD

**Discussion by abitrolly** [SCons/scons#4478] [Q&A]
How to catch and handle tool errors to provide users an actionable error message? 

While compiling Godot on Fedora I've got this error.
```make
[100%] Linking Program bin/godot.linuxbsd.editor.x86_64 ...
/usr/bin/ld: cannot find -lstdc++: No such file or directory
collect2: error: ld returned 1 exit status
scons: *** [bin/godot.linuxbsd.editor.x86_64] Error 1
scons: building terminated because of errors.
[Time elapsed: 01:07:12.255]
```

The cause is probably missing header library, but don't understand which. `libstdc++-devel-13.2.1-6.fc39.x86_64` is installed. Maybe the error is in something else. When I find the cause, and fix it manually, I want to add a check to the build, so that the next time this error occurs, the error message tells me exactly what to do. How to do that in SCons?

**UPDATE:** The cause of the error was [missing `libstdc++-static` lib](https://forum.godotengine.org/t/trying-to-build-godot-on-linux-linker-error-cant-link-to-std-library/39754/6?u=abitrolly). So how to tell that to users?

**Comment by mwichmann** [comment]
You can use the `Configure` functionality to both tune the build based on component availability, and to simply abort with a message if something is missing.  Have a glance here:

https://scons.org/doc/production/HTML/scons-user.html#chap-sconf

**Reply by abitrolly** [reply]
That looks like LBYL pattern - explicitly checking for every lib. Is it possible to implement EAFP instead? React and handle only specific errors?

**Reply by bdbaddog** [reply]
No.  There's no provisions for checking for a specific type of failure when you execute a builder to provide users a specific error message.

However as @mwichmann has sugguested, that's explicitly what Configure is meant for.
If you depend on a library, best to check if it exists before compiling your entire source tree

**Comment by mwichmann** [comment]
It's not clear what that would look like... in normal builds, SCons gets no information back from the external command except the exit code provided by the operating system, so it makes no attempt to decode except "it failed".  The `Configure` subsystem has special handling, it does collect the information back.  Those results are cached, so it doesn't redo a check unless there seems to be a reason to do so, or if you start from a fresh checkout. You can limit your configure checks to things that seem to be at risk, if you want.

This particular case is one, certainly - not everyone seems to feel as strongly as this: https://fedoraproject.org/wiki/C_and_C%2B%2B_v2#Static_Linking

**Reply by bdbaddog** [reply]
Or barring that, provide a FAQ, and/or display a link to FAQ if your build fails..
